### ISSUE TITLE
interrupt of long run: panic bug 

### ISSUE THREAD

**Issue by ldemailly** [fortio/fortio#119]
```
07:45:56 W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
07:45:56 W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
07:45:56 W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
07:45:56 W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
07:45:56 W http.go:667> Parsed non ok code 503 (HTTP/1.1 503)
07:48:33 I uihandler.go:426> UI: GET /fortio/?stop=Stop&runid=43 HTTP/1.1 216.239.45.130:55783 ()
07:48:33 C uihandler.go:153> Stop request from 216.239.45.130:55783 for 43
07:48:33 I periodic.go:499> T009 ended after 1h24m36.175139994s : 50762 calls. qps=10.000048973893362
07:48:33 I periodic.go:499> T005 ended after 1h24m36.175179648s : 50762 calls. qps=10.000048895775109
07:48:33 I periodic.go:499> T007 ended after 1h24m36.175199914s : 50762 calls. qps=10.000048855851155
07:48:33 I periodic.go:499> T006 ended after 1h24m36.17520971s : 50762 calls. qps=10.000048836553066
07:48:33 I periodic.go:499> T002 ended after 1h24m36.176164715s : 50762 calls. qps=10.000046955196641
07:48:33 I periodic.go:499> T008 ended after 1h24m36.176496053s : 50762 calls. qps=10.000046302462135
07:48:33 I periodic.go:499> T003 ended after 1h24m36.180292916s : 50762 calls. qps=10.000038822663623
07:48:33 I periodic.go:499> T000 ended after 1h24m36.181708568s : 50762 calls. qps=10.000036033840098
07:48:33 I periodic.go:499> T001 ended after 1h24m36.181749501s : 50762 calls. qps=10.000035953202428
07:48:33 I periodic.go:499> T004 ended after 1h24m36.184033277s : 50762 calls. qps=10.00003145418467
07:48:33 http: panic serving 172.10.234.41:51533: close of closed channel
goroutine 6673 [running]:
net/http.(*conn).serve.func1(0xc4202620a0)
        /usr/lib/go-1.8/src/net/http/server.go:1721 +0xd0
panic(0x8c0fc0, 0xc4205eeae0)
        /usr/lib/go-1.8/src/runtime/panic.go:489 +0x2cf
istio.io/fortio/periodic.(*RunnerOptions).Abort(0xc42025a780)
        /home/ldemailly/go/src/istio.io/fortio/periodic/periodic.go:233 +0x62
istio.io/fortio/fhttp.RunHTTPTest(0xc420012380, 0xc420072410, 0x0, 0x0)
        /home/ldemailly/go/src/istio.io/fortio/fhttp/httprunner.go:156 +0xdf5
istio.io/fortio/ui.Handler(0xb93400, 0xc4200122a0, 0xc4200d4500)
        /home/ldemailly/go/src/istio.io/fortio/ui/uihandler.go:294 +0xc12
net/http.HandlerFunc.ServeHTTP(0x976648, 0xb93400, 0xc4200122a0, 0xc4200d4500)
        /usr/lib/go-1.8/src/net/http/server.go:1942 +0x44
net/http.(*ServeMux).ServeHTTP(0xbd1b60, 0xb93400, 0xc4200122a0, 0xc4200d4500)
        /usr/lib/go-1.8/src/net/http/server.go:2238 +0x130
net/http.serverHandler.ServeHTTP(0xc4202380b0, 0xb93400, 0xc4200122a0, 0xc4200d4500)
        /usr/lib/go-1.8/src/net/http/server.go:2568 +0x92
net/http.(*conn).serve(0xc4202620a0, 0xb93c80, 0xc420672080)
        /usr/lib/go-1.8/src/net/http/server.go:1825 +0x612
created by net/http.(*Server).Serve
        /usr/lib/go-1.8/src/net/http/server.go:2668 +0x2ce
```

On page:
```
Up for 27h28m14.8s (since Wed Feb 7 02:55:42 2018).

1M 100qps ing-f2-f1 35.199.14.248/fortio2/fortio/fetch/echosrv1:8080/echo

Running load test... Results pending... 

Interrupt
Starting at 100 qps with 10 thread(s) [gomax 2] : exactly 1000000, 100000 calls each (total 1000000 + 0)
Ended after 1h24m36.184069079s : 507620 calls. qps=100
Sleep times : count 507620 avg 0.071686248 +/- 0.06425 min -2.920981178 max 0.099712334 sum 36389.3734
Aggregated Function Time : count 507620 avg 0.023398227 +/- 0.02961 min 7.4395e-05 max 3.02076883 sum 11877.4082
# range, mid point, percentile, count
>= 7.4395e-05 <= 0.001 , 0.000537198 , 0.00, 10
> 0.001 <= 0.002 , 0.0015 , 0.00, 8
> 0.002 <= 0.003 , 0.0025 , 0.02, 73
> 0.003 <= 0.004 , 0.0035 , 0.07, 265
> 0.004 <= 0.005 , 0.0045 , 0.12, 243
> 0.005 <= 0.006 , 0.0055 , 0.25, 659
> 0.006 <= 0.007 , 0.0065 , 0.39, 713
> 0.007 <= 0.008 , 0.0075 , 0.53, 710
> 0.008 <= 0.009 , 0.0085 , 0.93, 2028
> 0.009 <= 0.01 , 0.0095 , 2.05, 5688
> 0.01 <= 0.011 , 0.0105 , 4.30, 11449
> 0.011 <= 0.012 , 0.0115 , 7.85, 18020
> 0.012 <= 0.014 , 0.013 , 18.76, 55382
> 0.014 <= 0.016 , 0.015 , 33.19, 73246
> 0.016 <= 0.018 , 0.017 , 47.12, 70708
> 0.018 <= 0.02 , 0.019 , 57.92, 54796
> 0.02 <= 0.025 , 0.0225 , 76.14, 92513
> 0.025 <= 0.03 , 0.0275 , 85.62, 48098
> 0.03 <= 0.035 , 0.0325 , 90.28, 23693
> 0.035 <= 0.04 , 0.0375 , 92.90, 13265
> 0.04 <= 0.045 , 0.0425 , 94.45, 7904
> 0.045 <= 0.05 , 0.0475 , 95.56, 5623
> 0.05 <= 0.06 , 0.055 , 97.09, 7764
> 0.06 <= 0.07 , 0.065 , 98.02, 4691
> 0.07 <= 0.08 , 0.075 , 98.59, 2920
> 0.08 <= 0.09 , 0.085 , 98.96, 1889
> 0.09 <= 0.1 , 0.095 , 99.19, 1171
> 0.1 <= 0.12 , 0.11 , 99.44, 1266
> 0.12 <= 0.14 , 0.13 , 99.59, 769
> 0.14 <= 0.16 , 0.15 , 99.69, 498
> 0.16 <= 0.18 , 0.17 , 99.75, 279
> 0.18 <= 0.2 , 0.19 , 99.77, 134
> 0.2 <= 0.25 , 0.225 , 99.81, 183
> 0.25 <= 0.3 , 0.275 , 99.83, 92
> 0.3 <= 0.35 , 0.325 , 99.84, 80
> 0.35 <= 0.4 , 0.375 , 99.86, 56
> 0.4 <= 0.45 , 0.425 , 99.87, 53
> 0.45 <= 0.5 , 0.475 , 99.88, 82
> 0.5 <= 0.6 , 0.55 , 99.93, 239
> 0.6 <= 0.7 , 0.65 , 99.97, 221
> 0.7 <= 0.8 , 0.75 , 99.99, 111
> 0.8 <= 0.9 , 0.85 , 100.00, 12
> 0.9 <= 1 , 0.95 , 100.00, 1
> 1 <= 2 , 1.5 , 100.00, 9
> 3 <= 3.02077 , 3.01038 , 100.00, 6
# target 50% 0.0185332
# target 75% 0.0246867
# target 90% 0.0346953
# target 99% 0.0915867
# target 99.9% 0.538234
Code  -1 : 10 (0.0 %)
Code 200 : 501723 (98.8 %)
Code 503 : 5887 (1.2 %)
Response Header Sizes : count 507620 avg 164.01283 +/- 17.78 min 0 max 185 sum 83256194
Response Body/Total Sizes : count 507620 avg 166.43754 +/- 5.049 min 0 max 227 sum 84487022
```

**Comment by ldemailly** [comment]
same with a "until interrupted" run
Starting at 1000 qps with 8 thread(s) [gomax 2] until interrupted
```
07:56:28 I uihandler.go:426> UI: GET /fortio/?stop=Stop&runid=1 HTTP/1.1 216.239.45.130:52994 ()
07:56:28 C uihandler.go:153> Stop request from 216.239.45.130:52994 for 1
07:56:28 I periodic.go:499> T001 ended after 2.411858699s : 302 calls. qps=125.21463223579997
07:56:28 I periodic.go:499> T003 ended after 2.411891689s : 302 calls. qps=125.2129195425077
07:56:28 I periodic.go:499> T004 ended after 2.411917476s : 302 calls. qps=125.21158082939317
07:56:28 I periodic.go:499> T002 ended after 2.411882089s : 302 calls. qps=125.21341792675005
07:56:28 I periodic.go:499> T005 ended after 2.411927721s : 302 calls. qps=125.21104897570851
07:56:28 I periodic.go:499> T007 ended after 2.411975929s : 302 calls. qps=125.20854639092876
07:56:28 I periodic.go:499> T006 ended after 2.411940271s : 302 calls. qps=125.21039746759135
07:56:28 I periodic.go:499> T000 ended after 2.411933535s : 302 calls. qps=125.21074715269879
07:56:28 http: panic serving 216.239.45.130:42035: close of closed channel
goroutine 18 [running]:
net/http.(*conn).serve.func1(0xc420258000)
	/usr/lib/go-1.8/src/net/http/server.go:1721 +0xd0
panic(0x8c0fc0, 0xc4200151d0)
	/usr/lib/go-1.8/src/runtime/panic.go:489 +0x2cf
istio.io/fortio/periodic.(*RunnerOptions).Abort(0xc420075880)
	/home/ldemailly/go/src/istio.io/fortio/periodic/periodic.go:233 +0x62
istio.io/fortio/fhttp.RunHTTPTest(0xc4200128c0, 0xc420250680, 0x0, 0x0)
	/home/ldemailly/go/src/istio.io/fortio/fhttp/httprunner.go:156 +0xdf5
istio.io/fortio/ui.Handler(0xb93400, 0xc4200127e0, 0xc42000b900)
	/home/ldemailly/go/src/istio.io/fortio/ui/uihandler.go:294 +0xc12
net/http.HandlerFunc.ServeHTTP(0x976648, 0xb93400, 0xc4200127e0, 0xc42000b900)
	/usr/lib/go-1.8/src/net/http/server.go:1942 +0x44
net/http.(*ServeMux).ServeHTTP(0xbd1b60, 0xb93400, 0xc4200127e0, 0xc42000b900)
	/usr/lib/go-1.8/src/net/http/server.go:2238 +0x130
net/http.serverHandler.ServeHTTP(0xc42008f4a0, 0xb93400, 0xc4200127e0, 0xc42000b900)
	/usr/lib/go-1.8/src/net/http/server.go:2568 +0x92
net/http.(*conn).serve(0xc420258000, 0xb93c80, 0xc42024c300)
	/usr/lib/go-1.8/src/net/http/server.go:1825 +0x612
created by net/http.(*Server).Serve
	/usr/lib/go-1.8/src/net/http/server.go:2668 +0x2ce
```

**Comment by ldemailly** [comment]
simpler repro (1 c, talking to other server):

```
fortio server --loglevel debug
15:55:04 I logger.go:97> Log level is now 0 Debug (was 2 Info)
15:55:04 V stats.go:491> Will use [50 75 99 99.9] for percentiles
Https redirector running on :8081
UI starting - visit:
http://localhost:8080/fortio/
15:55:04 I uihandler.go:95> Guessing resources directory from runtime source location: true - /Users/ldemailly/go/src/istio.io/fortio/ui/uihandler.go
Fortio 0.6.8 grpc ping server listening on port 8079
Fortio 0.6.8 echo server listening on port :8080
15:55:28 I uihandler.go:425> UI: GET /fortio/?labels=Fortio&url=http%3A%2F%2Flocalhost%3A8085%2Fecho&qps=.1&t=&t=on&n=&c=1&p=50%2C+75%2C+99%2C+99.9&r=0.0001&H=User-Agent%3A+istio%2Ffortio-0.6.8&H=&H=&save=on&load=Start HTTP/1.1 [::1]:59728 ()
15:55:28 V uihandler.go:430> Header Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
15:55:28 V uihandler.go:430> Header Accept-Encoding: gzip, deflate, br
15:55:28 V uihandler.go:430> Header Accept-Language: en-US,en;q=0.9,fr;q=0.8
15:55:28 V uihandler.go:430> Header Cookie: last-serviceName=istio-ingress; grafana_sess=fcf952a9e1a61f2c; _ga=GA1.1.721844013.1495588107
15:55:28 V uihandler.go:430> Header Connection: keep-alive
15:55:28 V uihandler.go:430> Header Upgrade-Insecure-Requests: 1
15:55:28 V uihandler.go:430> Header User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.119 Safari/537.36
15:55:28 I uihandler.go:148> Starting load request from [::1]:59728 for http://localhost:8085/echo
15:55:28 V stats.go:491> Will use [50 75 99 99.9] for percentiles
15:55:28 V http.go:92> URLSchemeCheck &{URL:http://localhost:8085/echo NumConnections:1 Compression:false DisableFastClient:false HTTP10:false DisableKeepAlive:false AllowHalfClose:false initDone:true extraHeaders:map[User-Agent:[istio/fortio-0.6.8]] hostOverride: HTTPReqTimeOut:15s}
15:55:28 I uihandler.go:204> New run id 1
15:55:28 V periodic.go:188> WATCHER 1 First outstanding run starting, catching signal
15:55:28 V periodic.go:194> WATCHER 1 starting new watcher for signal! chan  g 0xc4202a0060 r 0xc4202ec180 (11)
15:55:28 V uihandler.go:274> adding header User-Agent: istio/fortio-0.6.8
15:55:28 I http.go:167> Setting regular extra header User-Agent: istio/fortio-0.6.8
15:55:28 D http.go:169> headers now map[User-Agent:[istio/fortio-0.6.8]]
15:55:28 I http.go:167> Setting regular extra header X-On-Behalf-Of: [::1]:59728
15:55:28 D http.go:169> headers now map[X-On-Behalf-Of:[[::1]:59728] User-Agent:[istio/fortio-0.6.8]]
15:55:28 I httprunner.go:72> Starting http test for http://localhost:8085/echo with 1 threads at 0.1 qps
15:55:28 V http.go:92> URLSchemeCheck &{URL:http://localhost:8085/echo NumConnections:1 Compression:false DisableFastClient:false HTTP10:false DisableKeepAlive:false AllowHalfClose:false initDone:true extraHeaders:map[User-Agent:[istio/fortio-0.6.8] X-On-Behalf-Of:[[::1]:59728]] hostOverride: HTTPReqTimeOut:15s}
15:55:28 D http.go:344> Using only the first of the addresses for localhost:8085 : [::1 127.0.0.1]
15:55:28 D http.go:346> Will go to ::1
15:55:28 D http.go:380> Created client:
{IP:::1 Port:8085 Zone:}
GET /echo HTTP/1.1
Host: localhost:8085
User-Agent: istio/fortio-0.6.8
X-On-Behalf-Of: [::1]:59728

15:55:28 D http.go:655> Read ok 116 total 116 so far (-0 headers = 116 data) HTTP/1.1 200 OK\r\nDate: Thu, 08 Feb 2018 23:55:28 GMT\r\nContent-Length: 0\r\nContent-Type: text/plain; charset=utf-8\r\n\r\n
15:55:28 D http.go:671> Code 200, looking for end of headers at 12 / 116, last CRLF 0
15:55:28 D http.go:692> headers are 116: HTTP/1.1 200 OK
Date: Thu, 08 Feb 2018 23:55:28 GMT
Content-Length: 0
Content-Type: text/plain; charset=utf-8
15:55:28 D http.go:708> found content length 0
15:55:28 V httprunner.go:98> first hit of url http://localhost:8085/echo: status 200, headers 116, total 116
HTTP/1.1 200 OK
Date: Thu, 08 Feb 2018 23:55:28 GMT
Content-Length: 0
Content-Type: text/plain; charset=utf-8
15:55:28 V periodic.go:356> Running single threaded
15:55:28 D httprunner.go:51> Calling in 0
15:55:28 D http.go:572> Reusing socket {{0xc420278380}}
15:55:28 D http.go:655> Read ok 116 total 116 so far (-0 headers = 116 data) HTTP/1.1 200 OK\r\nDate: Thu, 08 Feb 2018 23:55:28 GMT\r\nContent-Length: 0\r\nContent-Type: text/plain; charset=utf-8\r\n\r\n
15:55:28 D http.go:671> Code 200, looking for end of headers at 12 / 116, last CRLF 0
15:55:28 D http.go:692> headers are 116: HTTP/1.1 200 OK
Date: Thu, 08 Feb 2018 23:55:28 GMT
Content-Length: 0
Content-Type: text/plain; charset=utf-8
15:55:28 D http.go:708> found content length 0
15:55:28 D httprunner.go:54> Got in 200 hsz 116 sz 116
15:55:28 D periodic.go:477> T000 target next dur 10s - sleep 9.999720012s
15:55:32 I uihandler.go:425> UI: GET /fortio/?stop=Stop&runid=1 HTTP/1.1 [::1]:59733 ()
15:55:32 V uihandler.go:430> Header Connection: keep-alive
15:55:32 V uihandler.go:430> Header User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.119 Safari/537.36
15:55:32 V uihandler.go:430> Header Accept: */*
15:55:32 V uihandler.go:430> Header Referer: http://localhost:8080/fortio/?labels=Fortio&url=http%3A%2F%2Flocalhost%3A8085%2Fecho&qps=.1&t=&t=on&n=&c=1&p=50%2C+75%2C+99%2C+99.9&r=0.0001&H=User-Agent%3A+istio%2Ffortio-0.6.8&H=&H=&save=on&load=Start
15:55:32 V uihandler.go:430> Header Accept-Encoding: gzip, deflate, br
15:55:32 V uihandler.go:430> Header Accept-Language: en-US,en;q=0.9,fr;q=0.8
15:55:32 C uihandler.go:153> Stop request from [::1]:59733 for 1
15:55:32 V http.go:92> URLSchemeCheck &{URL:http://url.needed NumConnections:1 Compression:false DisableFastClient:false HTTP10:false DisableKeepAlive:false AllowHalfClose:false initDone:true extraHeaders:map[User-Agent:[istio/fortio-0.6.8]] hostOverride: HTTPReqTimeOut:15s}
15:55:32 V periodic.go:209> WATCHER 1 r.Stop readable
15:55:32 I periodic.go:499> T000 ended after 4.011791387s : 1 calls. qps=0.24926520437738806
15:55:32 V periodic.go:212> WATCHER 1 End of go routine
15:55:32 V periodic.go:216> WATCHER 1 Last watcher: resetting signal handler
15:55:32 V periodic.go:422> RUNNER r.Stop already closed
15:55:32 http: panic serving [::1]:59728: close of closed channel
goroutine 10 [running]:
net/http.(*conn).serve.func1(0xc42008d360)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:1697 +0xd0
panic(0x14819c0, 0x1572110)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/panic.go:491 +0x283
istio.io/fortio/periodic.(*RunnerOptions).Abort(0xc420278280)
	/Users/ldemailly/go/src/istio.io/fortio/periodic/periodic.go:233 +0x6b
istio.io/fortio/fhttp.RunHTTPTest(0xc4202f00e0, 0xc420294340, 0x0, 0x0)
	/Users/ldemailly/go/src/istio.io/fortio/fhttp/httprunner.go:156 +0x12e8
istio.io/fortio/ui.Handler(0x177a080, 0xc4202f0000, 0xc4202b8500)
	/Users/ldemailly/go/src/istio.io/fortio/ui/uihandler.go:294 +0xbaa
net/http.HandlerFunc.ServeHTTP(0x15362c8, 0x177a080, 0xc4202f0000, 0xc4202b8500)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:1918 +0x44
net/http.(*ServeMux).ServeHTTP(0x17c01a0, 0x177a080, 0xc4202f0000, 0xc4202b8500)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:2254 +0x130
net/http.serverHandler.ServeHTTP(0xc4202ba750, 0x177a080, 0xc4202f0000, 0xc4202b8500)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:2619 +0xb4
net/http.(*conn).serve(0xc42008d360, 0x177a880, 0xc4200111c0)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:1801 +0x71d
created by net/http.(*Server).Serve
	/usr/local/Cellar/go/1.9.2/libexec/src/net/http/server.go:2720 +0x288
```

**Comment by danehans** [comment]
Do you think the bug is due to [this limitation](https://github.com/istio/fortio/blob/master/periodic/periodic.go#L91-L94) of the Runner stop channel?

**Comment by ldemailly** [comment]
it's related but the uihandler is supposed to use that api correctly - I thought I had fixed that once for all in #64 but looks like I didn't

I just need ~1h of available quiet time to find/fix that bug but I have been swamped
